pkg/templates/accesstoresources/template.go

```diff
@@ -0,0 +1,206 @@
+package accesstoresources
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/pkg/errors"
+	"golang.stackrox.io/kube-linter/internal/stringutils"
+	"golang.stackrox.io/kube-linter/pkg/check"
+	"golang.stackrox.io/kube-linter/pkg/config"
+	"golang.stackrox.io/kube-linter/pkg/diagnostic"
+	"golang.stackrox.io/kube-linter/pkg/extract"
+	"golang.stackrox.io/kube-linter/pkg/lintcontext"
+	"golang.stackrox.io/kube-linter/pkg/objectkinds"
+	"golang.stackrox.io/kube-linter/pkg/templates"
+	"golang.stackrox.io/kube-linter/pkg/templates/accesstoresources/internal/params"
+	rbacV1 "k8s.io/api/rbac/v1"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+const (
+	templateKey = "access-to-resources"
+)
+
+var (
+	roleGVK               = rbacV1.SchemeGroupVersion.WithKind(objectkinds.Role)
+	clusterRoleGVK        = rbacV1.SchemeGroupVersion.WithKind(objectkinds.ClusterRole)
+	roleBindingGVK        = rbacV1.SchemeGroupVersion.WithKind(objectkinds.RoleBinding)
+	clusterRoleBindingGVK = rbacV1.SchemeGroupVersion.WithKind(objectkinds.ClusterRoleBinding)
+)
+
+func init() {
+	templates.Register(check.Template{
+		HumanName:   "Access to Resources",
+		Key:         templateKey,
+		Description: "Flag cluster role bindings and role bindings that grant access to the specified resource kinds and verbs",
+		SupportedObjectKinds: config.ObjectKindsDesc{
+			ObjectKinds: []string{
+				objectkinds.Role,
+				objectkinds.ClusterRole,
+				objectkinds.ClusterRoleBinding,
+				objectkinds.RoleBinding},
+		},
+		Parameters:             params.ParamDescs,
+		ParseAndValidateParams: params.ParseAndValidate,
+		Instantiate: params.WrapInstantiateFunc(func(p params.Params) (check.Func, error) {
+			resourceRegexes := make([]*regexp.Regexp, 0, len(p.Resources))
+			for _, res := range p.Resources {
+				r, err := regexp.Compile(res)
+				if err != nil {
+					return nil, errors.Wrapf(err, "invalid regex %s", res)
+				}
+				resourceRegexes = append(resourceRegexes, r)
+			}
+			verbRegexes := make([]*regexp.Regexp, 0, len(p.Verbs))
+			for _, verb := range p.Verbs {
+				v, err := regexp.Compile(verb)
+				if err != nil {
+					return nil, errors.Wrapf(err, "invalid regex %s", verb)
+				}
+				verbRegexes = append(verbRegexes, v)
+			}
+			return func(lintCtx lintcontext.LintContext, object lintcontext.Object) []diagnostic.Diagnostic {
+				bindingGVK := extract.GVK(object.K8sObject)
+				if bindingGVK == roleBindingGVK {
+					binding, ok := object.K8sObject.(*rbacV1.RoleBinding)
+					if !ok {
+						return nil
+					}
+					namespace := stringutils.OrDefault(binding.Namespace, "default")
+					return findRole(binding.RoleRef.Name, namespace, lintCtx, resourceRegexes, verbRegexes, p.FlagRolesNotFound)
+				}
```

Comment:
Unless I'm missing something, the outer `if` is unnecessary since this is already covered by the casting?
```suggestion
					binding, ok := object.K8sObject.(*rbacV1.RoleBinding)
					if ok {
					  namespace := stringutils.OrDefault(binding.Namespace, "default")
					  return findRole(binding.RoleRef.Name, namespace, lintCtx, resourceRegexes, verbRegexes, p.FlagRolesNotFound)
					}
```

Reply:
The object passed in to this function `func(lintCtx lintcontext.LintContext, object lintcontext.Object)` could be one of two kinds: RoleBinding or ClusterRoleBinding (see the checks that use this template `access-to-create-pods.yaml` and `access-to-secrets.yaml`).  The two kinds of objects have different schema and need to be processed differently. 

The`extract.GVK(object.K8sObject)` casts only the object's GVK, not the whole object.  Depending on the GVK, we cast the object to its kind and process accordingly in the outer `if`. For RoleBinding object, we will scan all the Role objects, and for ClusterRoleBinding all the ClusterRoles.